### PR TITLE
Document Symbols support

### LINKED ISSUE #382: Add support for `textDocument/documentSymbol`
Been enjoying my time so far with lexical! This is one of the features that I am direly missing.

I use this all the time in a file to jump to a function or symbol by name. This is one that I might possibly take a stab at being my first contribution.

### PR DESCRIPTION
**Pull request by scohen**
Added document symbols, which supports the following symbols:

  * Modules
  * Functions, both private and public
  * Typespecs
  * Module Attributes
  * ExUnit describe / setup / tests

Fixes #382

### CODE REVIEW

**Comment by scohen** [comment]
Typed the following:
```elixir
defprotocol Inspect do
end
```
...and got 
```
08:59:16.527 [error] ** (ErlangError) Erlang error: {:exception, :undef, [{Inspect, :inspect, [:line, %Inspect.Opts{base: :decimal, binaries: :infer, char_lists: :infer, charlists: :infer, custom_options: [], inspect_fun: &Inspect.inspect/2, limit: 50, pretty: false, printable_limit: 4096, safe: true, structs: true, syntax_colors: [], width: 80}], []}, {Kernel, :inspect, 2, [file: 'lib/kernel.ex', line: 2254]}, {LXSourceror.Range, :get_range_for_qualified_call_without_arguments, 1, [file: 'lib/sourceror/range.ex', line: 429]}, {LXSourceror.Range, :get_range, 2, [file: 'lib/sourceror/range.ex', line: 12]}, {LXSourceror.Range, :get_range_for_pair, 2, [file: 'lib/sourceror/range.ex', line: 549]}, {LXSourceror.Range, :get_range, 2, [file: 'lib/sourceror/range.ex', line: 12]}, {LXSourceror.Range, :get_range_for_pair, 2, [file: 'lib/sourceror/range.ex', line: 549]}, {LXSourceror.Range, :get_range, 2, [file: 'lib/sourceror/range.ex', line: 12]}]}
    (kernel 8.5.4.2) erpc.erl:702: :erpc.call/5
    (lx_server 0.5.0) lib/lexical/server/provider/handlers/document_symbols.ex:16: LXical.Server.Provider.Handlers.DocumentSymbols.handle/2
    (lx_server 0.5.0) lib/lexical/server/provider/queue.ex:99: anonymous fn/2 in LXical.Server.Provider.Queue.State.as_task/2
    (elixir 1.14.5) lib/task/supervised.ex:89: Task.Supervised.invoke_mfa/2
    (elixir 1.14.5) lib/task/supervised.ex:34: Task.Supervised.reply/4
    (stdlib 4.3.1.2) proc_lib.erl:240: :proc_lib.init_p_do_apply/3
```

**Comment by scohen** [comment]
@scohen can i please get a review on this?

**Comment by scottming** [comment]
Just got back from the countryside yesterday and experienced it for a couple of days,today I'll take a look.

**Review by scottming — COMMENTED** [review]
Another aspect that I think could enhance user experience is, do we really need to display the *full name* of `struct`? It seems redundant with modules, and I think using just the last name of the module is sufficient.

Lastly, why doesn't `defimpl` have symbols? like: `apps/protocol/lib/lexical/protocol/convertibles/lexical.document.edit.ex` It's also a module.

File: apps/server/lib/lexical/server/provider/handlers/document_symbols.ex [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+      range: root.range,
+      selection_range: root.detail_range
+    )
+  end
+
+  defp to_kind(:struct), do: :struct
+  defp to_kind(:module), do: :module
+  defp to_kind(:variable), do: :variable
+  defp to_kind(:public_function), do: :function
+  defp to_kind(:private_function), do: :function
```
[the remark this answers is not in the document]
Comment (scohen) [reply apps/server/lib/lexical/server/provider/handlers/document_symbols.ex:48]: my worry here is twofold. First, pedantically, these aren't methods, and the distinction between function and method isn't correct. Secondly, we'll need to remember the mapping between public / private functions and functions / methods, and be consistent whenever we display one of them (the symbol kind is reused a lot).

**Comment by scohen** [comment]
> Lastly, why doesn't defimpl have symbols? l

Because we don't have a defimpl extractor. Nor do we have a defprotocol extractor. Once those are added, we'll get symbols automatically.

**Comment by scohen** [comment]
@scottming this is holding up other stuff. I'm merging.

**Pull request merged**